test/new-e2e/system-probe/system-probe-test-env.go

```diff
@@ -90,84 +109,103 @@ func NewTestEnv(name, securityGroups, subnets, x86InstanceType, armInstanceType
+					shutdownRegisterArgs := command.Args{
+						Create: pulumi.Sprintf(
+							"shutdown -P +%.0f", opts.ShutdownPeriod.Minutes(),
+						),
+						Sudo: true,
+					}
+					shutdownRegisterDone, err := remoteRunner.Command("shutdown-"+instance.Arch, &shutdownRegisterArgs, pulumi.DependsOn(scenarioDone.Dependencies))
+					if err != nil {
+						return fmt.Errorf("failed to schedule shutdown: %w", err)
+					}
+					depends = []pulumi.Resource{shutdownRegisterDone}
+				} else {
+					depends = scenarioDone.Dependencies
 				}
-				shutdownRegisterDone, err := remoteRunner.Command("shutdown-"+instance.Arch, &shutdownRegisterArgs, pulumi.DependsOn(scenarioDone.Dependencies))
-				if err != nil {
-					return fmt.Errorf("failed to schedule shutdown: %w", err)
+
+				if opts.UploadDependencies {
+					// Copy dependencies to micro-vms. Directory '/opt/kernel-version-testing'
+					// is mounted to all micro-vms. Each micro-vm extract the context on boot.
+					filemanager := command.NewFileManager(remoteRunner)
+					_, err = filemanager.CopyFile(
+						fmt.Sprintf("%s/dependencies-%s.tar.gz", DD_AGENT_TESTING_DIR, instance.Arch),
+						fmt.Sprintf("/opt/kernel-version-testing/dependencies-%s.tar.gz", instance.Arch),
+						pulumi.DependsOn(depends),
+						pulumi.Provider(commandProvider),
+					)
+					if err != nil {
+						return fmt.Errorf("copy file: %w", err)
+					}
 				}
-				depends = []pulumi.Resource{shutdownRegisterDone}
-			} else {
-				depends = scenarioDone.Dependencies
 			}
 
-			if opts.UploadDependencies {
-				// Copy dependencies to micro-vms. Directory '/opt/kernel-version-testing'
-				// is mounted to all micro-vms. Each micro-vm extract the context on boot.
-				filemanager := command.NewFileManager(remoteRunner)
-				_, err = filemanager.CopyFile(
-					fmt.Sprintf("%s/dependencies-%s.tar.gz", DD_AGENT_TESTING_DIR, instance.Arch),
-					fmt.Sprintf("/opt/kernel-version-testing/dependencies-%s.tar.gz", instance.Arch),
-					pulumi.DependsOn(depends),
-				)
-				if err != nil {
-					return fmt.Errorf("copy file: %w", err)
-				}
+			return nil
+		}, opts.FailOnMissing)
+		// Only retry if we failed to dial libvirt.
+		// Libvirt daemon on the server occassionally crashes with the following error
+		// "End of file while reading data: Input/output error"
+		// The root cause of this is unknown. The problem usually fixes itself upon retry.
+		if err != nil {
+			if strings.Contains(fmt.Sprintf("%s", err), "failed to dial libvirt") {
+				fmt.Printf("[Error] Failed to dial libvirt. Retrying stack.")
+				return retry.RetryableError(err)
+			} else {
+				return err
 			}
 		}
 
 		return nil
-	}, opts.FailOnMissing)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create stack: %w", err)
+	}); retryErr != nil {
+		return nil, fmt.Errorf("failed to create stack: %w", retryErr)
 	}
 
-	b, err := json.MarshalIndent(upResult.Summary, "", "	")
+	err = outputsToFile(upResult.Outputs)
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
This should be handled by a deserialiser - as soon as we document it, good for now